styx-scheduler-service/src/main/java/com/spotify/styx/BackfillTriggerManager.java

```diff
@@ -110,53 +124,76 @@ private void triggerBackfill(Backfill backfill, Map<String, Long> backfillStates
 
     final Workflow workflow = workflowOpt.get();
 
+    while (true) {
+      try {
+        if (!storage.runInTransaction(tx -> tryTriggerAndProgress(tx, backfill.id(), workflow))) {
+          break;
+        }
+      } catch (IOException e) {
+        // transaction failure, yield
+        LOG.debug("transaction failure when working on backfill {}", backfill, e);
+        break;
+      }
+    }
+  }
+
+  private Boolean tryTriggerAndProgress(StorageTransaction tx,
+                                        String id, Workflow workflow) {
+    final Backfill momentBackfill =
+        storage.backfill(id).orElseThrow(RuntimeException::new);
```

Comment:
One remote issue could be that instance A gets very slow right after this line, with read `nextTrigger` = 1, while a faster instance B proceeds with the backfill bringing `nextTrigger` to, for example, 5. WFI 1 in the meantime finishes with SUCCESS. Afterwards, instance A resumes and triggers WFI 1 _again_ and set `nextTrigger` to 2, causing the whole backfill to be re-triggered.

Perhaps `storage.backfill(id)` should be in the same transaction as `tx.store(<Backfill>)` that appears later?
If that is the case we would still trigger partition 1 twice, but at least we wouldn't regress `nextCounter` to 2.

Reply:
It may trigger twice in that extreme case but wouldn’t be able to progress next trigger because transaction would fail that.

Reply:
Shouldn't this be `tx.backfill()` for the transaction to fail instead of `storage.backfill()`?

Reply:
The slow node wouldn’t be able to store a stale backfill if it has been updated by the other node.

Reply:
As we figured out we need to read the backfill in the beginning of the transaction.